<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Provides an infrastructure for efficient processing of large-scale genetic and phenotypic data including core functions for: 1) fitting linear mixed models, 2) constructing marker-based genomic relationship matrices, 3) estimating genetic parameters (heritability and correlation), 4) performing genomic prediction and genetic risk profiling, and 5) single or multi-marker association analyses.
    Rohde et al. (2019) &lt;doi:10.1101/503631&gt;.">
<title>Statistical Tools for Quantitative Genetic Analyses • qgg</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Statistical Tools for Quantitative Genetic Analyses">
<meta property="og:description" content="Provides an infrastructure for efficient processing of large-scale genetic and phenotypic data including core functions for: 1) fitting linear mixed models, 2) constructing marker-based genomic relationship matrices, 3) estimating genetic parameters (heritability and correlation), 4) performing genomic prediction and genetic risk profiling, and 5) single or multi-marker association analyses.
    Rohde et al. (2019) &lt;doi:10.1101/503631&gt;.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">qgg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/psoerensen/qgg/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level3">
<h3 id="an-r-package-for-quantitative-genetic-and-genomic-analyses">An R package for Quantitative Genetic and Genomic analyses<a class="anchor" aria-label="anchor" href="#an-r-package-for-quantitative-genetic-and-genomic-analyses"></a>
</h3>
<p>The <strong>qgg</strong> package was developed based on the hypothesis that certain regions on the genome, so-called <em>genomic features</em>, may be enriched for causal variants affecting the trait. Several genomic feature classes can be formed based on previous studies and different sources of information such as genes, chromosomes or biological pathways.</p>
<p><strong>qgg</strong> provides an infrastructure for efficient processing of large-scale genetic and phenotypic data including core functions for:</p>
<ul>
<li>fitting linear mixed models<br>
</li>
<li>construction of genomic relationship matrices<br>
</li>
<li>estimating genetic parameters (heritability and correlation)<br>
</li>
<li>genomic prediction<br>
</li>
<li>single marker association analysis<br>
</li>
<li>gene set enrichment analysis</li>
</ul>
<p><strong>qgg</strong> handles large-scale data by taking advantage of:</p>
<ul>
<li>multi-core processing using <a href="https://www.openmp.org/" class="external-link">openMP</a><br>
</li>
<li>multithreaded matrix operations implemented in BLAS libraries (e.g. <a href="https://www.openblas.net/" class="external-link">OpenBLAS</a>, <a href="http://math-atlas.sourceforge.net/" class="external-link">ATLAS</a> or <a href="https://software.intel.com/en-us/mkl" class="external-link">MKL</a>)<br>
</li>
<li>fast and memory-efficient batch processing of genotype data stored in binary files (e.g. <a href="https://www.cog-genomics.org/plink2" class="external-link">PLINK</a> bedfiles)</li>
</ul>
<p>The <strong>qgg</strong> package provides a range of genomic feature modeling approaches, including genomic feature best linear unbiased prediction (GFBLUP) models, implemented using likelihood or Bayesian methods. Multiple features and multiple traits can be included in these models and different genetic models (e.g. additive, dominance, gene by gene and gene by environment interactions) can be used. Further extensions include a weighted GFBLUP model using differential weighting of the individual genetic marker relationships. Marker set tests, which are computationally very fast, can be performed. These marker set tests allow the rapid analyses of different layers of genomic feature classes to discover genomic features potentially enriched for causal variants. Marker set tests can thus facilitate more accurate prediction models.</p>
</div>
<div class="section level3">
<h3 id="install">Install<a class="anchor" aria-label="anchor" href="#install"></a>
</h3>
<p>You can install qgg from CRAN with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"qgg"</span><span class="op">)</span></span></code></pre></div>
<p>The most recent version of <code>qgg</code> can be obtained from github:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"psoerensen/qgg"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="tutorials">Tutorials<a class="anchor" aria-label="anchor" href="#tutorials"></a>
</h3>
<p>Below is a set of tutorials used for the qgg package:</p>
<p>This tutorial provides a brief introduction to R package qgg using small simulated data examples.<br><a href="https://psoerensen.github.io/qgtutorials/Quick-tutorials-for-qgg-package.pdf" class="external-link">Practicals_brief_introduction</a></p>
<p>This tutorial provides an introduction to R package qgg using 1000G data.<br><a href="https://psoerensen.github.io/qgtutorials/1000G-tutorials-for-qgg-package.pdf" class="external-link">Practicals_1000G_tutorials</a></p>
<p>This tutorial provide a simple introduction to polygenic risk scoring (PRS) of complex traits and diseases using simulated data. The practical will be a mix of theoretical and practical exercises in R that are used for illustrating/applying the theory presented in the corresponding lecture notes on polygenic risk scoring.<br><a href="https://psoerensen.github.io/qgtutorials/Practicals_human_example.pdf" class="external-link">Practicals_human_example</a></p>
<p>In this tutorial we will be analysing quantitative traits observed in a mice population. The mouse data consist of phenotypes for traits related to growth and obesity (e.g. body weight, glucose levels in blood), pedigree information, and genetic marker data.<br><a href="https://psoerensen.github.io/qgtutorials/Practicals_mouse_example.pdf" class="external-link">Practicals_mouse_example</a></p>
</div>
<div class="section level3">
<h3 id="notes">Notes<a class="anchor" aria-label="anchor" href="#notes"></a>
</h3>
<p>Below is a set of notes for the quantitative genetic theory, statistical models and methods implemented in the qgg package:</p>
<p><a href="https://psoerensen.github.io/qgnotes/Quantitative-Genetics-Theory.pdf" class="external-link">Quantitative Genetics Theory</a></p>
<p><a href="https://psoerensen.github.io/qgnotes/Estimation-of-Genetic-Predisposition.pdf" class="external-link">Estimation of Genetic Predisposition</a></p>
<p><a href="https://psoerensen.github.io/qgnotes/Estimation-of-Genetic-Parameters.pdf" class="external-link">Estimation of Genetic Parameters</a></p>
<p><a href="https://psoerensen.github.io/qgnotes/LMM.pdf" class="external-link">Linear Mixed Models</a></p>
<p><a href="https://psoerensen.github.io/qgnotes/BLUP.pdf" class="external-link">Best Linear Unbiased Prediction Models</a></p>
<p><a href="https://psoerensen.github.io/qgnotes/REML.pdf" class="external-link">REstricted Maximum Likelihood Methods</a></p>
<p><a href="https://psoerensen.github.io/qgnotes/GSEA.pdf" class="external-link">Gene Set Enrichment Analysis</a></p>
<p><a href="https://psoerensen.github.io/qgnotes/BLR.pdf" class="external-link">Bayesian Linear Regression Models</a></p>
<div class="section level4">
<h4 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Edwards SM, Thomsen B, Madsen P, Sørensen P. 2015. Partitioning of genomic variance reveals biological pathways associated with udder health and milk production traits in dairy cattle. <em>Genet Sel Evol</em> 47:60. <a href="doi:10.1186/s12711-015-0132-6" class="uri">doi:10.1186/s12711-015-0132-6</a><br>
</li>
<li>Edwards SM, Sørensen IF, Sarup P, Mackay TFC, Sørensen P. 2016. Genomic prediction for quantitative traits is improved by mapping variants to gene ontology categories in <em>Drosophila melanogaster</em>. <em>Genetics</em> 203:1871–1883. <a href="doi:10.1534/genetics.116.187161" class="uri">doi:10.1534/genetics.116.187161</a>
</li>
<li>Ehsani A, Janss L, Pomp D, Sørensen P. 2015. Decomposing genomic variance using information from GWA, GWE and eQTL analysis. <em>Anim Genet</em> 47:165–173. <a href="doi:10.1111/age.12396" class="uri">doi:10.1111/age.12396</a>
</li>
<li>Fang L, Sahana G, Ma P, Su G, Yu Y, Zhang S, Lund MS, Sørensen P. 2017. Exploring the genetic architecture and improving genomic prediction accuracy for mastitis and milk production traits in dairy cattle by mapping variants to hepatic transcriptomic regions responsive to intra-mammary infection. <em>Genet Sel Evol</em> 49:1–18. <a href="doi:10.1186/s12711-017-0319-0" class="uri">doi:10.1186/s12711-017-0319-0</a>
</li>
<li>Fang L, Sahana G, Su G, Yu Y, Zhang S, Lund MS, Sørensen P. 2017. Integrating sequence-based GWAS and RNA-seq provides novel insights into the genetic basis of mastitis and milk production in dairy cattle. <em>Sci Rep</em> 7:45560. <a href="doi:10.1038/srep45560" class="uri">doi:10.1038/srep45560</a>
</li>
<li>Fang L, Sørensen P, Sahana G, Panitz F, Su G, Zhang S, Yu Y, Li B, Ma L, Liu G, Lund MS, Thomsen B. 2018. MicroRNA-guided prioritization of genome-wide association signals reveals the importance of microRNA-target gene networks for complex traits in cattle. <em>Sci Rep</em> 8:1–14. <a href="doi:10.1038/s41598-018-27729-y" class="uri">doi:10.1038/s41598-018-27729-y</a>
</li>
<li>Ørsted M, Rohde PD, Hoffmann AA, Sørensen P, Kristensen TN. 2017. Environmental variation partitioned into separate heritable components. <em>Evolution</em> (N Y) 72:136–152. <a href="doi:10.1111/evo.13391" class="uri">doi:10.1111/evo.13391</a>
</li>
<li>Ørsted M, Hoffmann AA, Rohde PD, Sørensen P, Kristensen TN. 2018. Strong impact of thermal environment on the quantitative genetic basis of a key stress tolerance trait. <em>Heredity</em> (Edinb). <a href="doi:10.1038/s41437-018-0117-7" class="uri">doi:10.1038/s41437-018-0117-7</a>
</li>
<li>Rohde PD, Krag K, Loeschcke V, Overgaard J, Sørensen P, Kristensen TN. 2016. A quantitative genomic approach for analysis of fitness and stress related traits in a <em>Drosophila melanogaster model population</em>. <em>Int J Genomics</em> 2016:1–11.</li>
<li>Rohde PD, Demontis D, Cuyabano BCD, The GEMS Group, Børglum AD, Sørensen P. 2016. Covariance Association Test (CVAT) identify genetic markers associated with schizophrenia in functionally associated biological processes. <em>Genetics</em> 203:1901–1913. <a href="doi:10.1534/genetics.116.189498" class="uri">doi:10.1534/genetics.116.189498</a>
</li>
<li>Rohde PD, Gaertner B, Ward K, Sørensen P, Mackay TFC. 2017. Genomic analysis of genotype-by-social environment interaction for <em>Drosophila melanogaster</em>. <em>Genetics</em> 206:1969–1984. <a href="doi:10.1534/genetics.117.200642/-/DC1.1" class="uri">doi:10.1534/genetics.117.200642/-/DC1.1</a>
</li>
<li>Rohde PD, Østergaard S, Kristensen TN, Sørensen P, Loeschcke V, Mackay TFC, Sarup P. 2018. Functional validation of candidate genes detected by genomic feature models. <em>G3 Genes, Genomes, Genet</em> 8:1659–1668. <a href="doi:10.1534/g3.118.200082" class="uri">doi:10.1534/g3.118.200082</a>
</li>
<li>Sarup P, Jensen J, Ostersen T, Henryon M, Sørensen P. 2016. Increased prediction accuracy using a genomic feature model including prior information on quantitative trait locus regions in purebred Danish Duroc pigs. <em>BMC Genet</em> 17:11. <a href="doi:10.1186/s12863-015-0322-9" class="uri">doi:10.1186/s12863-015-0322-9</a>
</li>
<li>Sørensen P, de los Campos G, Morgante F, Mackay TFC, Sorensen D. 2015. Genetic control of environmental variation of two quantitative traits of <em>Drosophila melanogaster</em> revealed by whole-genome sequencing. <em>Genetics</em> 201:487–497. <a href="doi:10.1534/genetics.115.180273" class="uri">doi:10.1534/genetics.115.180273</a>
</li>
<li>Sørensen IF, Edwards SM, Rohde PD, Sørensen P. 2017. Multiple trait covariance association test identifies gene ontology categories associated with chill coma recovery time in <em>Drosophila melanogaster</em>. <em>Sci Rep</em> 7:2413. <a href="doi:10.1038/s41598-017-02281-3" class="uri">doi:10.1038/s41598-017-02281-3</a>
</li>
</ol>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=qgg" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/psoerensen/qgg/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/psoerensen/qgg/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing qgg</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Peter Soerensen <br><small class="roles"> Author, maintainer </small>  </li>
<li>Palle Duun Rohde <br><small class="roles"> Author </small>  </li>
<li>Izel Fourie Soerensen <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Peter Soerensen, Palle Duun Rohde, Izel Fourie Soerensen.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
