<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Quality control is a fundamental step in GWAS summary statistics analysis.
The function is equipped to handle various tasks including mapping marker ids,
checking the effect allele and its frequency, determining build versions, and 
excluding data based on multiple criteria."><title>Quality Control of Marker Summary Statistics — checkStat • qgg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quality Control of Marker Summary Statistics — checkStat"><meta property="og:description" content="Quality control is a fundamental step in GWAS summary statistics analysis.
The function is equipped to handle various tasks including mapping marker ids,
checking the effect allele and its frequency, determining build versions, and 
excluding data based on multiple criteria."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">qgg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/psoerensen/qgg/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Quality Control of Marker Summary Statistics</h1>
      <small class="dont-index">Source: <a href="https://github.com/psoerensen/qgg/blob/HEAD/R/genomic_statistics.R" class="external-link"><code>R/genomic_statistics.R</code></a></small>
      <div class="d-none name"><code>checkStat.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Quality control is a fundamental step in GWAS summary statistics analysis.
The function is equipped to handle various tasks including mapping marker ids,
checking the effect allele and its frequency, determining build versions, and 
excluding data based on multiple criteria.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">checkStat</span><span class="op">(</span></span>
<span>  Glist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  stat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  excludeMAF <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  excludeMAFDIFF <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  excludeINFO <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  excludeCGAT <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  excludeINDEL <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  excludeDUPS <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  excludeMHC <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  excludeMISS <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  excludeHWE <span class="op">=</span> <span class="fl">1e-12</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>Glist</dt>
<dd><p>List containing information about genotype matrix stored on disk.</p></dd>


<dt>stat</dt>
<dd><p>Data frame of marker summary statistics. It should either follow 
the "internal" or "external" format.</p></dd>


<dt>excludeMAF</dt>
<dd><p>Numeric. Exclusion threshold for minor allele frequency. Default is 0.01.</p></dd>


<dt>excludeMAFDIFF</dt>
<dd><p>Numeric. Threshold for excluding markers based on allele frequency difference. Default is 0.05.</p></dd>


<dt>excludeINFO</dt>
<dd><p>Numeric. Exclusion threshold for info score. Default is 0.8.</p></dd>


<dt>excludeCGAT</dt>
<dd><p>Logical. Exclude ambiguous alleles (CG or AT). Default is TRUE.</p></dd>


<dt>excludeINDEL</dt>
<dd><p>Logical. Exclude insertion/deletion markers. Default is TRUE.</p></dd>


<dt>excludeDUPS</dt>
<dd><p>Logical. Exclude markers with duplicated ids. Default is TRUE.</p></dd>


<dt>excludeMHC</dt>
<dd><p>Logical. Exclude markers located in MHC region. Default is FALSE.</p></dd>


<dt>excludeMISS</dt>
<dd><p>Numeric. Exclusion threshold for sample missingness. Default is 0.05.</p></dd>


<dt>excludeHWE</dt>
<dd><p>Numeric. Exclusion threshold for Hardy Weinberg Equilibrium test p-value. Default is 1e-12.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data frame with processed and quality-controlled summary statistics.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Performs quality control on GWAS summary statistics, which includes:
- Mapping marker ids to LD reference panel data.
- Checking effect allele, frequency, and build version.
- Excluding based on various criteria like MAF, HWE, INDELS, and more.</p>

<p>The function works with both "internal" and "external" formats of summary statistics. 
When the summary statistics format is "external", the function maps marker ids based 
on chr-pos-ref-alt information. It also aligns the effect allele with the LD reference 
panel and flips effect sizes if necessary. When allele frequencies are not provided, 
it uses the frequencies from the genotype data.</p>
<p>Required headers for external summary statistics:
 marker, chr, pos, ea, nea, eaf, b, seb, stat, p, n</p>
<p>Required headers for internal summary statistics:
 rsids, chr, pos, ea, nea, eaf, b, seb, stat, p, n</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Peter Soerensen</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Peter Soerensen, Palle Duun Rohde, Izel Fourie Soerensen.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

