<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="In the Bayesian multiple regression model, the posterior density of the model parameters depends
on the likelihood of the data given the parameters and a prior probability for the model parameters.
The choice of the prior for marker effects can influence the type and extent of shrinkage induced in the model."><title>Finemapping using Bayesian Linear Regression Models — gmap • qgg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Finemapping using Bayesian Linear Regression Models — gmap"><meta property="og:description" content="In the Bayesian multiple regression model, the posterior density of the model parameters depends
on the likelihood of the data given the parameters and a prior probability for the model parameters.
The choice of the prior for marker effects can influence the type and extent of shrinkage induced in the model."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">qgg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/psoerensen/qgg/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Finemapping using Bayesian Linear Regression Models</h1>
      <small class="dont-index">Source: <a href="https://github.com/psoerensen/qgg/blob/HEAD/R/genomic_bayes.R" class="external-link"><code>R/genomic_bayes.R</code></a></small>
      <div class="d-none name"><code>gmap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>In the Bayesian multiple regression model, the posterior density of the model parameters depends
on the likelihood of the data given the parameters and a prior probability for the model parameters.
The choice of the prior for marker effects can influence the type and extent of shrinkage induced in the model.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gmap</span><span class="op">(</span></span>
<span>  y <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  X <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  W <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  stat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  trait <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Glist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rsids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  b <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bm <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mask <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  LD <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  vg <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  vb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ve <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ssg_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ssb_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sse_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lambda <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaleY <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  shrinkLD <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  shrinkCor <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  formatLD <span class="op">=</span> <span class="st">"dense"</span>,</span>
<span>  pruneLD <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  r2 <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  checkLD <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  h2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pi <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  updateB <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  updateG <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  updateE <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  updatePi <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  adjustE <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  models <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  checkConvergence <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  critVe <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  critVg <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  critVb <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  critPi <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  ntrial <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  nug <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  nub <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  nue <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  msize <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ve_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  vg_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  nit <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  nburn <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  nit_local <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nit_global <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"bayesC"</span>,</span>
<span>  algorithm <span class="op">=</span> <span class="st">"mcmc"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>y</dt>
<dd><p>A vector or matrix of phenotypes.</p></dd>


<dt>X</dt>
<dd><p>A matrix of covariates.</p></dd>


<dt>W</dt>
<dd><p>A matrix of centered and scaled genotypes.</p></dd>


<dt>stat</dt>
<dd><p>Dataframe with marker summary statistics.</p></dd>


<dt>trait</dt>
<dd><p>Integer used for selection traits in covs object.</p></dd>


<dt>sets</dt>
<dd><p>A list of character vectors where each vector represents a set of items. If the names
of the sets are not provided, they are named as "Set1", "Set2", etc.</p></dd>


<dt>fit</dt>
<dd><p>List of results from gbayes.</p></dd>


<dt>Glist</dt>
<dd><p>List of information about genotype matrix stored on disk.</p></dd>


<dt>chr</dt>
<dd><p>Chromosome for which to fit BLR models.</p></dd>


<dt>rsids</dt>
<dd><p>Character vector of rsids.</p></dd>


<dt>ids</dt>
<dd><p>vector of individuals used in the study</p></dd>


<dt>b</dt>
<dd><p>Vector or matrix of marginal marker effects.</p></dd>


<dt>bm</dt>
<dd><p>Vector or matrix of adjusted marker effects for the BLR model.</p></dd>


<dt>seb</dt>
<dd><p>Vector or matrix of standard error of marginal effects.</p></dd>


<dt>mask</dt>
<dd><p>Vector or matrix specifying if marker should be ignored.</p></dd>


<dt>LD</dt>
<dd><p>List with sparse LD matrices.</p></dd>


<dt>n</dt>
<dd><p>Scalar or vector of number of observations for each trait.</p></dd>


<dt>vg</dt>
<dd><p>Scalar or matrix of genetic (co)variances.</p></dd>


<dt>vb</dt>
<dd><p>Scalar or matrix of marker (co)variances.</p></dd>


<dt>ve</dt>
<dd><p>Scalar or matrix of residual (co)variances.</p></dd>


<dt>ssg_prior</dt>
<dd><p>Scalar or matrix of prior genetic (co)variances.</p></dd>


<dt>ssb_prior</dt>
<dd><p>Scalar or matrix of prior marker (co)variances.</p></dd>


<dt>sse_prior</dt>
<dd><p>Scalar or matrix of prior residual (co)variances.</p></dd>


<dt>lambda</dt>
<dd><p>Vector or matrix of lambda values</p></dd>


<dt>scaleY</dt>
<dd><p>Logical indicating if y should be scaled.</p></dd>


<dt>shrinkLD</dt>
<dd><p>Logical indicating if LD should be shrunk.</p></dd>


<dt>shrinkCor</dt>
<dd><p>Logical indicating if cor should be shrunk.</p></dd>


<dt>formatLD</dt>
<dd><p>Character specifying LD format (default is "dense").</p></dd>


<dt>pruneLD</dt>
<dd><p>Logical indicating if LD pruning should be applied.</p></dd>


<dt>r2</dt>
<dd><p>Scalar providing value for r2 threshold used in pruning</p></dd>


<dt>checkLD</dt>
<dd><p>Logical indicating if LD matches summary statistics.</p></dd>


<dt>h2</dt>
<dd><p>Trait heritability.</p></dd>


<dt>pi</dt>
<dd><p>Proportion of markers in each marker variance class.</p></dd>


<dt>updateB</dt>
<dd><p>Logical indicating if marker (co)variances should be updated.</p></dd>


<dt>updateG</dt>
<dd><p>Logical indicating if genetic (co)variances should be updated.</p></dd>


<dt>updateE</dt>
<dd><p>Logical indicating if residual (co)variances should be updated.</p></dd>


<dt>updatePi</dt>
<dd><p>Logical indicating if pi should be updated.</p></dd>


<dt>adjustE</dt>
<dd><p>Logical indicating if residual variance should be adjusted.</p></dd>


<dt>models</dt>
<dd><p>List structure with models evaluated in bayesC.</p></dd>


<dt>checkConvergence</dt>
<dd><p>Logical indicating if convergences should be checked.</p></dd>


<dt>critVe</dt>
<dd><p>Scalar providing value for z-score threshold used in checking convergence for Ve</p></dd>


<dt>critVg</dt>
<dd><p>Scalar providing value for z-score threshold used in checking convergence for Vg</p></dd>


<dt>critVb</dt>
<dd><p>Scalar providing value for z-score threshold used in checking convergence for Vg</p></dd>


<dt>critPi</dt>
<dd><p>Scalar providing value for z-score threshold used in checking convergence for Pi</p></dd>


<dt>ntrial</dt>
<dd><p>Integer providing number of trials used if convergence is not obtaines</p></dd>


<dt>nug</dt>
<dd><p>Scalar or vector of prior degrees of freedom for genetic (co)variances.</p></dd>


<dt>nub</dt>
<dd><p>Scalar or vector of prior degrees of freedom for marker (co)variances.</p></dd>


<dt>nue</dt>
<dd><p>Scalar or vector of prior degrees of freedom for residual (co)variances.</p></dd>


<dt>verbose</dt>
<dd><p>Logical; if TRUE, it prints more details during iteration.</p></dd>


<dt>msize</dt>
<dd><p>Integer providing number of markers used in computation of sparseld</p></dd>


<dt>threshold</dt>
<dd><p>Scalar providing value for threshold used in adjustment of B</p></dd>


<dt>ve_prior</dt>
<dd><p>Scalar or matrix of prior residual (co)variances.</p></dd>


<dt>vg_prior</dt>
<dd><p>Scalar or matrix of prior genetic (co)variances.</p></dd>


<dt>tol</dt>
<dd><p>Convergence criteria used in gbayes.</p></dd>


<dt>nit</dt>
<dd><p>Number of iterations.</p></dd>


<dt>nburn</dt>
<dd><p>Number of burnin iterations.</p></dd>


<dt>nit_local</dt>
<dd><p>Number of local iterations.</p></dd>


<dt>nit_global</dt>
<dd><p>Number of global iterations.</p></dd>


<dt>method</dt>
<dd><p>Method used (e.g. "bayesN","bayesA","bayesL","bayesC","bayesR").</p></dd>


<dt>algorithm</dt>
<dd><p>Specifies the algorithm.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list containing:</p><ul><li><p><code>bm</code>Vector or matrix of posterior means for marker effects.</p></li>
<li><p><code>dm</code>Vector or matrix of posterior means for marker inclusion probabilities.</p></li>
<li><p><code>vb</code>Scalar or vector of posterior means for marker variances.</p></li>
<li><p><code>vg</code>Scalar or vector of posterior means for genomic variances.</p></li>
<li><p><code>ve</code>Scalar or vector of posterior means for residual variances.</p></li>
<li><p><code>rb</code>Matrix of posterior means for marker correlations.</p></li>
<li><p><code>rg</code>Matrix of posterior means for genomic correlations.</p></li>
<li><p><code>re</code>Matrix of posterior means for residual correlations.</p></li>
<li><p><code>pi</code>Vector of posterior probabilities for models.</p></li>
<li><p><code>h2</code>Vector of posterior means for model probability.</p></li>
<li><p><code>param</code>List of current parameters used for restarting the analysis.</p></li>
<li><p><code>stat</code>Matrix of marker information and effects used for genomic risk scoring.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function implements Bayesian linear regression models to provide unified mapping of 
genetic variants, estimate genetic parameters (e.g. heritability), and predict disease risk.
It is designed to handle various genetic architectures and scale efficiently with large datasets.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Peter Sørensen</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span></span>
<span><span class="co"># Plink bed/bim/fam files</span></span>
<span><span class="va">bedfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sample_chr"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="st">".bed"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"qgg"</span><span class="op">)</span></span>
<span><span class="va">bimfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sample_chr"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="st">".bim"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"qgg"</span><span class="op">)</span></span>
<span><span class="va">famfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sample_chr"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="st">".fam"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"qgg"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Prepare Glist</span></span>
<span><span class="va">Glist</span> <span class="op">&lt;-</span> <span class="fu"><a href="gprep.html">gprep</a></span><span class="op">(</span>study<span class="op">=</span><span class="st">"Example"</span>, bedfiles<span class="op">=</span><span class="va">bedfiles</span>, bimfiles<span class="op">=</span><span class="va">bimfiles</span>, famfiles<span class="op">=</span><span class="va">famfiles</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulate phenotype</span></span>
<span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="gsim.html">gsim</a></span><span class="op">(</span>Glist<span class="op">=</span><span class="va">Glist</span>, chr<span class="op">=</span><span class="fl">1</span>, nt<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute single marker summary statistics</span></span>
<span><span class="va">stat</span> <span class="op">&lt;-</span> <span class="fu"><a href="glma.html">glma</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">sim</span><span class="op">$</span><span class="va">y</span>, Glist<span class="op">=</span><span class="va">Glist</span>, scale<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">stat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define fine-mapping regions </span></span>
<span><span class="va">sets</span> <span class="op">&lt;-</span> <span class="va">Glist</span><span class="op">$</span><span class="va">rsids</span></span>
<span><span class="va">Glist</span><span class="op">$</span><span class="va">chr</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"21"</span>,<span class="st">"1"</span>,<span class="va">Glist</span><span class="op">$</span><span class="va">chr</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> </span>
<span><span class="va">Glist</span><span class="op">$</span><span class="va">chr</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"22"</span>,<span class="st">"2"</span>,<span class="va">Glist</span><span class="op">$</span><span class="va">chr</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Fine map</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">gmap</span><span class="op">(</span>Glist<span class="op">=</span><span class="va">Glist</span>, stat<span class="op">=</span><span class="va">stat</span>, sets<span class="op">=</span><span class="va">sets</span>, verbose<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>            method<span class="op">=</span><span class="st">"bayesC"</span>, nit<span class="op">=</span><span class="fl">1500</span>, nburn<span class="op">=</span><span class="fl">500</span>, pi<span class="op">=</span><span class="fl">0.001</span><span class="op">)</span></span>
<span>            </span>
<span><span class="va">fit</span><span class="op">$</span><span class="va">post</span>  <span class="co"># Posterior inference for every fine-mapped region</span></span>
<span><span class="va">fit</span><span class="op">$</span><span class="va">conv</span>  <span class="co"># Convergence statistics for every fine-mapped region</span></span>
<span></span>
<span><span class="co"># Posterior inference for marker effect</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">stat</span><span class="op">)</span>             </span>
<span></span></code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Peter Soerensen, Palle Duun Rohde, Izel Fourie Soerensen.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

