<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="All functions in qgg relies on a simple data infrastructure that takes five main input sources;
phenotype data (y), covariate data (X), genotype data (G or Glist), a genomic relationship
matrix (GRM or GRMlist) and genetic marker sets (sets).
The genotypes are stored in a matrix (n x m (individuals x markers)) in memory (G) or in a
binary file on disk (Glist).
It is only for small data sets that the genotype matrix (G) can stored in memory. For large data
sets the genotype matrix has to stored in a binary file on disk (Glist). Glist is as a list
structure that contains information about the genotypes in the binary file.
The gprep function prepares the Glist, and is required for downstream analyses of large-scale
genetic data. Typically, the Glist is prepared once, and saved as an *.Rdata-file.
The gprep function reads genotype information from binary PLINK files, and creates the Glist
object that contains general information about the genotypes such as reference alleles,
allele frequencies and missing genotypes, and construct a binary file on the disk that contains
the genotypes as allele counts of the alternative allele (memory usage = (n x m)/4 bytes).
The gprep function can also be used to prepare sparse ld matrices.
The r2 metric used is the pairwise correlation between markers (allele count alternative allele)
in a specified region of the genome. The marker genotype is allele count of the alternative allele
which is assumed to be centered and scaled.
The Glist structure is used as input parameter for a number of qgg core functions including:
1) construction of genomic relationship matrices (grm), 2) construction of sparse ld matrices,
3) estimating genomic parameters (greml), 4) single marker association analyses (glma),
5) gene set enrichment analyses (gsea), and 6) genomic prediction from genotypes
and phenotypes (gsolve) or genotypes and summary statistics (gscore)."><title>Prepare genotype data for all statistical analyses — gprep • qgg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Prepare genotype data for all statistical analyses — gprep"><meta property="og:description" content="All functions in qgg relies on a simple data infrastructure that takes five main input sources;
phenotype data (y), covariate data (X), genotype data (G or Glist), a genomic relationship
matrix (GRM or GRMlist) and genetic marker sets (sets).
The genotypes are stored in a matrix (n x m (individuals x markers)) in memory (G) or in a
binary file on disk (Glist).
It is only for small data sets that the genotype matrix (G) can stored in memory. For large data
sets the genotype matrix has to stored in a binary file on disk (Glist). Glist is as a list
structure that contains information about the genotypes in the binary file.
The gprep function prepares the Glist, and is required for downstream analyses of large-scale
genetic data. Typically, the Glist is prepared once, and saved as an *.Rdata-file.
The gprep function reads genotype information from binary PLINK files, and creates the Glist
object that contains general information about the genotypes such as reference alleles,
allele frequencies and missing genotypes, and construct a binary file on the disk that contains
the genotypes as allele counts of the alternative allele (memory usage = (n x m)/4 bytes).
The gprep function can also be used to prepare sparse ld matrices.
The r2 metric used is the pairwise correlation between markers (allele count alternative allele)
in a specified region of the genome. The marker genotype is allele count of the alternative allele
which is assumed to be centered and scaled.
The Glist structure is used as input parameter for a number of qgg core functions including:
1) construction of genomic relationship matrices (grm), 2) construction of sparse ld matrices,
3) estimating genomic parameters (greml), 4) single marker association analyses (glma),
5) gene set enrichment analyses (gsea), and 6) genomic prediction from genotypes
and phenotypes (gsolve) or genotypes and summary statistics (gscore)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">qgg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/psoerensen/qgg/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Prepare genotype data for all statistical analyses</h1>
      <small class="dont-index">Source: <a href="https://github.com/psoerensen/qgg/blob/HEAD/R/genomic_matrix.R" class="external-link"><code>R/genomic_matrix.R</code></a></small>
      <div class="d-none name"><code>gprep.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>All functions in qgg relies on a simple data infrastructure that takes five main input sources;
phenotype data (y), covariate data (X), genotype data (G or Glist), a genomic relationship
matrix (GRM or GRMlist) and genetic marker sets (sets).</p>
<p>The genotypes are stored in a matrix (n x m (individuals x markers)) in memory (G) or in a
binary file on disk (Glist).</p>
<p>It is only for small data sets that the genotype matrix (G) can stored in memory. For large data
sets the genotype matrix has to stored in a binary file on disk (Glist). Glist is as a list
structure that contains information about the genotypes in the binary file.</p>
<p>The gprep function prepares the Glist, and is required for downstream analyses of large-scale
genetic data. Typically, the Glist is prepared once, and saved as an *.Rdata-file.</p>
<p>The gprep function reads genotype information from binary PLINK files, and creates the Glist
object that contains general information about the genotypes such as reference alleles,
allele frequencies and missing genotypes, and construct a binary file on the disk that contains
the genotypes as allele counts of the alternative allele (memory usage = (n x m)/4 bytes).</p>
<p>The gprep function can also be used to prepare sparse ld matrices.
The r2 metric used is the pairwise correlation between markers (allele count alternative allele)
in a specified region of the genome. The marker genotype is allele count of the alternative allele
which is assumed to be centered and scaled.</p>
<p>The Glist structure is used as input parameter for a number of qgg core functions including:
1) construction of genomic relationship matrices (grm), 2) construction of sparse ld matrices,
3) estimating genomic parameters (greml), 4) single marker association analyses (glma),
5) gene set enrichment analyses (gsea), and 6) genomic prediction from genotypes
and phenotypes (gsolve) or genotypes and summary statistics (gscore).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gprep</span><span class="op">(</span></span>
<span>  Glist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  task <span class="op">=</span> <span class="st">"prepare"</span>,</span>
<span>  study <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fnBED <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ldfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bedfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bimfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  famfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mapfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rsids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  assembly <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  msize <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  r2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  kb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cm <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>Glist</dt>
<dd><p>A list containing information about the genotype matrix stored on disk.</p></dd>


<dt>task</dt>
<dd><p>A character string specifying the task to perform. Possible tasks are "prepare" (default), "sparseld", "ldscores", "ldsets", and "geneticmap".</p></dd>


<dt>study</dt>
<dd><p>The name of the study.</p></dd>


<dt>fnBED</dt>
<dd><p>Path and filename of the .bed binary file used to store genotypes on disk.</p></dd>


<dt>ldfiles</dt>
<dd><p>Path and filename of the .ld binary files used for storing the sparse LD matrix on disk.</p></dd>


<dt>bedfiles</dt>
<dd><p>A vector of filenames for the PLINK bed-files.</p></dd>


<dt>bimfiles</dt>
<dd><p>A vector of filenames for the PLINK bim-files.</p></dd>


<dt>famfiles</dt>
<dd><p>A vector of filenames for the PLINK fam-files.</p></dd>


<dt>mapfiles</dt>
<dd><p>A vector of filenames for the mapfiles.</p></dd>


<dt>ids</dt>
<dd><p>A vector of individual identifiers used in the study.</p></dd>


<dt>rsids</dt>
<dd><p>A vector of marker rsids used in the study.</p></dd>


<dt>assembly</dt>
<dd><p>Character string indicating the name of the assembly.</p></dd>


<dt>overwrite</dt>
<dd><p>A logical value; if TRUE, the binary genotype/LD file will be overwritten.</p></dd>


<dt>msize</dt>
<dd><p>Number of markers used in the computation of sparseld.</p></dd>


<dt>r2</dt>
<dd><p>A threshold value (more context might be beneficial, e.g., threshold for what?).</p></dd>


<dt>kb</dt>
<dd><p>Size of the genomic region in kilobases (kb).</p></dd>


<dt>cm</dt>
<dd><p>Size of the genomic region in centimorgans (cm).</p></dd>


<dt>ncores</dt>
<dd><p>Number of processing cores to be used for genotype processing.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Returns a list structure (Glist) with information about the genotypes.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Peter Soerensen</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span></span>
<span><span class="va">bedfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sample_chr1.bed"</span>, package <span class="op">=</span> <span class="st">"qgg"</span><span class="op">)</span></span>
<span><span class="va">bimfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sample_chr1.bim"</span>, package <span class="op">=</span> <span class="st">"qgg"</span><span class="op">)</span></span>
<span><span class="va">famfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sample_chr1.fam"</span>, package <span class="op">=</span> <span class="st">"qgg"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Glist</span> <span class="op">&lt;-</span> <span class="fu">gprep</span><span class="op">(</span>study<span class="op">=</span><span class="st">"Example"</span>, bedfiles<span class="op">=</span><span class="va">bedfiles</span>, bimfiles<span class="op">=</span><span class="va">bimfiles</span>,</span>
<span>             famfiles<span class="op">=</span><span class="va">famfiles</span><span class="op">)</span></span>
<span></span></code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Peter Soerensen, Palle Duun Rohde, Izel Fourie Soerensen.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

