<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Computes genomic predictions using single marker summary statistics and observed genotypes."><title>Genomic scoring based on single marker summary statistics — gscore • qgg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Genomic scoring based on single marker summary statistics — gscore"><meta property="og:description" content="Computes genomic predictions using single marker summary statistics and observed genotypes."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">qgg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/psoerensen/qgg/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Genomic scoring based on single marker summary statistics</h1>
      <small class="dont-index">Source: <a href="https://github.com/psoerensen/qgg/blob/HEAD/R/genomic_score.R" class="external-link"><code>R/genomic_score.R</code></a></small>
      <div class="d-none name"><code>gscore.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Computes genomic predictions using single marker summary statistics and observed genotypes.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gscore</span><span class="op">(</span></span>
<span>  Glist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bedfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bimfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  famfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  stat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaleMarker <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scaleGRS <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  impute <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  msize <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>Glist</dt>
<dd><p>List of information about genotype matrix. Default is NULL.</p></dd>


<dt>chr</dt>
<dd><p>Chromosome for which genomic scores is computed. Default is NULL.</p></dd>


<dt>bedfiles</dt>
<dd><p>Names of the PLINK bed-files. Default is NULL.</p></dd>


<dt>bimfiles</dt>
<dd><p>Names of the PLINK bim-files. Default is NULL.</p></dd>


<dt>famfiles</dt>
<dd><p>Names of the PLINK fam-files. Default is NULL.</p></dd>


<dt>stat</dt>
<dd><p>Matrix of single marker effects. Default is NULL.</p></dd>


<dt>fit</dt>
<dd><p>Fit object output from gbayes. Default is NULL.</p></dd>


<dt>ids</dt>
<dd><p>Vector of individuals used in the analysis. Default is NULL.</p></dd>


<dt>scaleMarker</dt>
<dd><p>Logical; if TRUE the genotype markers are scaled to mean zero and variance one. Default is TRUE.</p></dd>


<dt>scaleGRS</dt>
<dd><p>Logical; if TRUE the GRS are scaled to mean zero and variance one. Default is TRUE.</p></dd>


<dt>impute</dt>
<dd><p>Logical; if TRUE, missing genotypes are set to its expected value (2*af where af is allele frequency). Default is TRUE.</p></dd>


<dt>msize</dt>
<dd><p>Number of genotype markers used for batch processing. Default is 100.</p></dd>


<dt>ncores</dt>
<dd><p>Number of cores used in the analysis. Default is 1.</p></dd>


<dt>verbose</dt>
<dd><p>Logical; if TRUE, more details are printed during optimization. Default is FALSE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Returns the genomic scores based on the provided parameters.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Peter Soerensen</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span></span>
<span> <span class="co">## Plink bed/bim/fam files</span></span>
<span> <span class="va">bedfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sample_chr"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="st">".bed"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"qgg"</span><span class="op">)</span></span>
<span> <span class="va">bimfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sample_chr"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="st">".bim"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"qgg"</span><span class="op">)</span></span>
<span> <span class="va">famfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sample_chr"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="st">".fam"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"qgg"</span><span class="op">)</span></span>
<span> </span>
<span> <span class="co"># Summarize bed/bim/fam files</span></span>
<span> <span class="va">Glist</span> <span class="op">&lt;-</span> <span class="fu"><a href="gprep.html">gprep</a></span><span class="op">(</span>study<span class="op">=</span><span class="st">"Example"</span>, bedfiles<span class="op">=</span><span class="va">bedfiles</span>, bimfiles<span class="op">=</span><span class="va">bimfiles</span>, famfiles<span class="op">=</span><span class="va">famfiles</span><span class="op">)</span></span>
<span> </span>
<span> <span class="co"># Simulate phenotype</span></span>
<span> <span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="gsim.html">gsim</a></span><span class="op">(</span>Glist<span class="op">=</span><span class="va">Glist</span>, chr<span class="op">=</span><span class="fl">1</span>, nt<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span> </span>
<span> <span class="co"># Compute single marker summary statistics</span></span>
<span> <span class="va">stat</span> <span class="op">&lt;-</span> <span class="fu"><a href="glma.html">glma</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">sim</span><span class="op">$</span><span class="va">y</span>, Glist<span class="op">=</span><span class="va">Glist</span>, scale<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span> </span>
<span> <span class="co"># Compute genomic scores</span></span>
<span> <span class="va">gsc</span> <span class="op">&lt;-</span> <span class="fu">gscore</span><span class="op">(</span>Glist <span class="op">=</span> <span class="va">Glist</span>, stat <span class="op">=</span> <span class="va">stat</span><span class="op">)</span></span>
<span> </span></code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Peter Soerensen, Palle Duun Rohde, Izel Fourie Soerensen.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

