<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The ldsc function is used for LDSC analysis"><title>LD score regression — ldsc • qgg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="LD score regression — ldsc"><meta property="og:description" content="The ldsc function is used for LDSC analysis"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">qgg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/psoerensen/qgg/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>LD score regression</h1>
      <small class="dont-index">Source: <a href="https://github.com/psoerensen/qgg/blob/HEAD/R/genomic_correlation.R" class="external-link"><code>R/genomic_correlation.R</code></a></small>
      <div class="d-none name"><code>ldsc.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The ldsc function is used for LDSC analysis</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ldsc</span><span class="op">(</span></span>
<span>  Glist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ldscores <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  z <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  b <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  af <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  stat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  intercept <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  what <span class="op">=</span> <span class="st">"h2"</span>,</span>
<span>  SE.h2 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  SE.rg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  blk <span class="op">=</span> <span class="fl">200</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>Glist</dt>
<dd><p>list of information about genotype matrix stored on disk</p></dd>


<dt>ldscores</dt>
<dd><p>vector of LD scores (optional as LD scores are stored within Glist)</p></dd>


<dt>z</dt>
<dd><p>matrix of z statistics for n traits</p></dd>


<dt>b</dt>
<dd><p>matrix of marker effects for n traits if z matrix not is given</p></dd>


<dt>seb</dt>
<dd><p>matrix of standard errors of marker effects for n traits if z matrix not is given</p></dd>


<dt>af</dt>
<dd><p>vector of allele frequencies</p></dd>


<dt>stat</dt>
<dd><p>dataframe with marker summary statistics</p></dd>


<dt>n</dt>
<dd><p>vector of sample sizes for the traits (element i corresponds to column vector i in z matrix)</p></dd>


<dt>intercept</dt>
<dd><p>logical if TRUE the LD score regression includes intercept</p></dd>


<dt>what</dt>
<dd><p>either computation of heritability (what="h2") or genetic correlation between traits (what="rg")</p></dd>


<dt>SE.h2</dt>
<dd><p>logical if TRUE standard errors and significance for the heritability estimates are computed using a block jackknife approach</p></dd>


<dt>SE.rg</dt>
<dd><p>logical if TRUE standard errors and significance for the genetic correlations are computed using a block jackknife approach</p></dd>


<dt>blk</dt>
<dd><p>numeric size of the blocks used in the jackknife estimation of standard error (default = 200)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Returns a matrix of heritability estimates when what="h2", and if SE.h2=TRUE standard errors (SE) and significance levels (P) are returned. 
        If what="rg" an n-by-n matrix of correlations is returned where the diagonal elements being h2 estimates. 
        If SE.rg=TRUE a list is returned with n-by-n matrices of genetic correlations, estimated standard errors and significance levels.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Peter Soerensen</p>
<p>Palle Duun Rohde</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span></span>
<span></span>
<span><span class="co"># Plink bed/bim/fam files</span></span>
<span> <span class="co">#bedfiles &lt;- system.file("extdata", paste0("sample_chr",1:2,".bed"), package = "qgg")</span></span>
<span> <span class="co">#bimfiles &lt;- system.file("extdata", paste0("sample_chr",1:2,".bim"), package = "qgg")</span></span>
<span> <span class="co">#famfiles &lt;- system.file("extdata", paste0("sample_chr",1:2,".fam"), package = "qgg")</span></span>
<span> <span class="co">#</span></span>
<span> <span class="co">## Summarize bed/bim/fam files</span></span>
<span> <span class="co">#Glist &lt;- gprep(study="Example", bedfiles=bedfiles, bimfiles=bimfiles, famfiles=famfiles)</span></span>
<span></span>
<span> <span class="co">#</span></span>
<span> <span class="co">## Filter rsids based on MAF, missingness, HWE</span></span>
<span> <span class="co">#rsids &lt;-  gfilter(Glist = Glist, excludeMAF=0.05, excludeMISS=0.05, excludeHWE=1e-12) </span></span>
<span> <span class="co">#</span></span>
<span> <span class="co">## Compute sparse LD (msize=size of LD window)</span></span>
<span> <span class="co">##ldfiles &lt;- system.file("extdata", paste0("sample_chr",1:2,".ld"), package = "qgg")</span></span>
<span> <span class="co">##Glist &lt;- gprep(Glist, task="sparseld", msize=200, rsids=rsids, ldfiles=ldfiles, overwrite=TRUE)</span></span>
<span> <span class="co">#</span></span>
<span> <span class="co">#</span></span>
<span> <span class="co">##Simulate data</span></span>
<span> <span class="co">#W1 &lt;- getG(Glist, chr=1, scale=TRUE)</span></span>
<span> <span class="co">#W2 &lt;- getG(Glist, chr=2, scale=TRUE)</span></span>
<span></span>
<span> <span class="co">#W &lt;- cbind(W1,W2)</span></span>
<span> <span class="co">#causal &lt;- sample(1:ncol(W),5)</span></span>
<span></span>
<span> <span class="co">#b1 &lt;- rnorm(length(causal))</span></span>
<span> <span class="co">#b2 &lt;- rnorm(length(causal))</span></span>
<span> <span class="co">#y1 &lt;- W[, causal]%*%b1 + rnorm(nrow(W))</span></span>
<span> <span class="co">#y2 &lt;- W[, causal]%*%b2 + rnorm(nrow(W))</span></span>
<span></span>
<span> <span class="co">#data1 &lt;- data.frame(y = y1, mu = 1)</span></span>
<span> <span class="co">#data2 &lt;- data.frame(y = y2, mu = 1)</span></span>
<span> <span class="co">#X1 &lt;- model.matrix(y ~ 0 + mu, data = data1)</span></span>
<span> <span class="co">#X2 &lt;- model.matrix(y ~ 0 + mu, data = data2)</span></span>
<span></span>
<span> <span class="co">## Linear model analyses and single marker association test</span></span>
<span> <span class="co">#maLM1 &lt;- lma(y=y1, X=X1,W = W)</span></span>
<span> <span class="co">#maLM2 &lt;- lma(y=y2,X=X2,W = W)</span></span>
<span> <span class="co">#</span></span>
<span> <span class="co">## Compute heritability and genetic correlations for trait 1 and 2</span></span>
<span> <span class="co">#z1 &lt;- maLM1[,"stat"]</span></span>
<span> <span class="co">#z2 &lt;- maLM2[,"stat"]</span></span>
<span></span>
<span> <span class="co">#z &lt;- cbind(z1=z1,z2=z2)</span></span>
<span></span>
<span> <span class="co">#h2 &lt;- ldsc(Glist, z=z, n=c(500,500), what="h2")</span></span>
<span> <span class="co">#rg &lt;- ldsc(Glist, z=z, n=c(500,500), what="rg")</span></span>
<span></span>
<span></span>
<span></span></code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Peter Soerensen, Palle Duun Rohde, Izel Fourie Soerensen.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

